server/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -830,4 +864,27 @@ public void setPipelinesHaveRun() {
     public boolean pipelinesHaveRun() {
         return pipelinesHaveRun;
     }
+
+    /**
+     * Adds the pipeline to the list of executed pipelines, if listExecutedPipelines is true
+     * @param pipeline
+     */
+    public void addPipeline(String pipeline) {
+        if (listExecutedPipelines) {
+            if (executedPipelines == null) {
+                executedPipelines = new ArrayList<>();
+            }
+            executedPipelines.add(pipeline);
+        }
+    }
+
+    public List<String> getExecutedPipelines() {
```

Comment:
This'll want a docstring about the null case.

Reply:
Done.